### PR TITLE
[Asset Store] Add animation, viewpoint, dimensions, object type and license filters

### PR DESCRIPTION
**Pull request by D8H**
[Test in live in Storybook](http://gdevelop-storybook.s3-website-us-east-1.amazonaws.com/asset-store/tag-filter/latest/index.html?path=/story/assetstore-newobjectdialog--default)

I'm undecided about class names. I would like to avoid confusion between classes from the view and the ones from the model.

### CODE REVIEW

File: newIDE/app/src/AssetStore/AssetStoreFilterPanel.js
Comment (D8H) [inline newIDE/app/src/AssetStore/AssetStoreFilterPanel.js:171]: I'm not sure what is the bet way to add some margin on both sides of the slider.

**Comment by Bouh** [comment]
- [ ] Replace categories by **Filter** ? image

File: newIDE/app/src/AssetStore/AssetStoreFilterPanel.js
Comment (ClementPasteau) [inline newIDE/app/src/AssetStore/AssetStoreFilterPanel.js:58]: can tags not have a label?

File: newIDE/app/src/AssetStore/AssetStoreFilterPanel.js
Comment (ClementPasteau) [reply newIDE/app/src/AssetStore/AssetStoreFilterPanel.js:174]: I'd suggest adding custom style to the slider, the same way we've done for other material ui components.
(search in the code `makeStyles()`)
I would add some padding left / right to the `root` https://v4.mui.com/api/slider/#css)

File: newIDE/app/src/AssetStore/AssetStoreFilterPanel.js
Comment (ClementPasteau) [inline newIDE/app/src/AssetStore/AssetStoreFilterPanel.js:206]: this can probably just be `onChange=setColor`

File: newIDE/app/src/AssetStore/AssetStoreSearchFilter.js
Comment (ClementPasteau) [inline newIDE/app/src/AssetStore/AssetStoreSearchFilter.js:121]: Why can't you use the remainder operation `%` ?

File: newIDE/app/src/AssetStore/HexColorField.js
Comment (ClementPasteau) [inline newIDE/app/src/AssetStore/HexColorField.js:47]: Why not use a functional component? (especially for something we just created)

File: newIDE/app/src/AssetStore/HexColorField.js
Comment (ClementPasteau) [inline newIDE/app/src/AssetStore/HexColorField.js:95]: log here

File: newIDE/app/src/UI/EmptyPlaceholder.js
Comment (ClementPasteau) [inline newIDE/app/src/UI/EmptyPlaceholder.js:56]: Can we also refactor the name of the action to be `onAction` rather than `onAdd`?
Then the action is either "add" by default, or something else if another icon is provided

File: newIDE/app/src/UI/Search/UseSearchItem.js
Comment (ClementPasteau) [inline newIDE/app/src/UI/Search/UseSearchItem.js:39]: is this to avoid multiple calls when selecting multiple filters quickly?

File: newIDE/app/src/AssetStore/AssetStoreContext.js
Comment (D8H) [reply newIDE/app/src/AssetStore/AssetStoreContext.js:178]: The setter is called in AssetStoreFilterPanel when the user change a criteria value.

File: newIDE/app/src/AssetStore/AssetStoreSearchFilter.js
Comment (D8H) [reply newIDE/app/src/AssetStore/AssetStoreSearchFilter.js:121]: I also prefer the formula with the double `%` but it's not a lot more simple and this method was used in the engine so I kept the same implementation.

File: newIDE/app/src/AssetStore/HexColorField.js
Comment (D8H) [reply newIDE/app/src/AssetStore/HexColorField.js:47]: It's a copy of an existing component and I wasn't sure it would be kept.

File: newIDE/app/src/AssetStore/AssetStoreContext.js
Comment (4ian) [reply newIDE/app/src/AssetStore/AssetStoreContext.js:178]: Calling a setState function (`setSearchFilters`) seems weird though.

Ensuring that `currentFilters` stays stable across render if the items inside have not changed is a great thing :) 
But the way to do it should be to do a `useMemo` (which looks a bit weird to be honest):

```
const currentFilters = React.useMemo(() => [
     animatedFilter,
    viewpointFilter,
    dimensionFilter,
    objectTypeFilter,
    colorFilter,
    licenseFilter,
    ], [
     animatedFilter,
    viewpointFilter,
    dimensionFilter,
    objectTypeFilter,
    colorFilter,
    licenseFilter,
    ])
```

which will properly return always the same array (so you get the stability you need) if dependencies (the second array) are always the same.

File: newIDE/app/src/AssetStore/AssetStoreContext.js
Comment (4ian) [inline newIDE/app/src/AssetStore/AssetStoreContext.js:154]: This object will be a new one at each render, do you need to `useMemo` on it? 
Otherwise it's passed to `const assetStoreState = React.useMemo(` and will make it change at each render even if the filters did not change?

File: newIDE/app/src/AssetStore/AssetStoreContext.js
Comment (4ian) [reply newIDE/app/src/AssetStore/AssetStoreContext.js:178]: EDIT: I may have not grasped entirely the problem, but this state does not looks like a state and so we should avoid it, I think useMemo is better.

File: newIDE/app/src/AssetStore/HexColorField.js
Comment (ClementPasteau) [reply newIDE/app/src/AssetStore/HexColorField.js:54]: I feel like we should keep it, like you mentioned, the usage is different, so we may end up modifying it in the future without affecting the other (and vice versa)

File: newIDE/app/src/UI/EmptyPlaceholder.js
Comment (ClementPasteau) [reply newIDE/app/src/UI/EmptyPlaceholder.js:56]: weird, maybe just because VSCode is lost, it's only used 2 or 3 times across the app though, so the refactor could be done manually

File: newIDE/app/src/AssetStore/AssetStoreFilterPanel.js
Comment (4ian) [inline newIDE/app/src/AssetStore/AssetStoreFilterPanel.js:52]: Can we remove margins for this (`noMargin`), or is the AccordionBody not adding margins and so we need it?

File: newIDE/app/src/UI/Search/UseSearchItem.js
Comment (ClementPasteau) [reply newIDE/app/src/UI/Search/UseSearchItem.js:39]: oh right, I misread the line for "filter for efficiency" instead of "to .filter() for efficiency" :)

File: newIDE/app/src/AssetStore/AssetStoreFilterPanel.js
Comment (4ian) [reply newIDE/app/src/AssetStore/AssetStoreFilterPanel.js:172]: I would recommend to go for a 
`<Column><Line noMargin><Slider ...></Line></Column>`

Comment (4ian) [reply newIDE/app/src/AssetStore/AssetStoreFilterPanel.js:172]: A slider by default, like any component, takes all the width available, which is good. It's the parent job to add layouting with margins (i.e: Line, Column)

**Review by ClementPasteau — COMMENTED** [review]
Great work overall 🌮 
lgtm as it's quite maintainable and easy to read.
A few things to cleanup, maybe in the next PR
I'll let @4ian review it too

File: newIDE/app/src/AssetStore/AssetStoreSearchFilter.js
Comment (4ian) [inline newIDE/app/src/AssetStore/AssetStoreSearchFilter.js:142]:
```suggestion
      // Hue is not relevant.
```

File: newIDE/app/src/AssetStore/HexColorField.js
Comment (4ian) [reply newIDE/app/src/AssetStore/HexColorField.js:54]: Do you think a prop like `preferHexFormat={true}` would work to reuse ColorField?

File: newIDE/app/src/UI/Search/UseSearchItem.js
Comment (4ian) [inline newIDE/app/src/UI/Search/UseSearchItem.js:11]: Is this class really needed? It seems to only have one subclass?

Comment (D8H) [reply newIDE/app/src/UI/Search/UseSearchItem.js:11]: It's to keep the logic about filtering on tags in this file. It might help to refactor when the old filters will be dropped or merged with the new filters.

File: newIDE/app/src/AssetStore/AssetStoreFilterPanel.js
Comment (D8H) [reply newIDE/app/src/AssetStore/AssetStoreFilterPanel.js:52]: It looks like AccordionBody not adding margins. Without it, the content is not aligned with the title.

File: newIDE/app/src/AssetStore/HexColorField.js
Comment (D8H) [reply newIDE/app/src/AssetStore/HexColorField.js]: There are already a lot of color conversions, it might make it hard to understand.

File: newIDE/app/src/AssetStore/AssetStoreSearchFilter.js
Comment (D8H) [reply newIDE/app/src/AssetStore/AssetStoreSearchFilter.js:115]: `-14 % 8 = -6` and `-14 mod % 8 = 2`, but I guess Clement meant a formula like `(x % y + x) % y` to avoid a call to floor.
It's true that it can't be negative but I think a mod make it more explicit.
uh, I copied pasted a documentation that is a bit confusing.

File: newIDE/app/src/AssetStore/AssetStoreContext.js
Comment (D8H) [reply newIDE/app/src/AssetStore/AssetStoreContext.js:164]: I made a change. It seems cleaner but I'm not sure if there is something else to change.

Initially, I did it by analogy to `useFilters()`, so it probably have the same issue. It create a new instance each call.

File: newIDE/app/src/AssetStore/AssetStoreFilterPanel.js
Comment (4ian) [reply newIDE/app/src/AssetStore/AssetStoreFilterPanel.js:52]: Ok 👍

File: newIDE/app/src/AssetStore/HexColorField.js
Comment (4ian) [inline newIDE/app/src/AssetStore/HexColorField.js:70]:
```suggestion
  // It keeps the inputed text if the color has not changed.
```

**Review by 4ian — APPROVED** [review]
LGTM after a few additional comments :)

**Pull request merged**
